### DISCUSSION TITLE
Calculation of ERA

### DISCUSSION THREAD

**Discussion by watercs** [groops-devs/groops#212] [Q&A]
There may be a implicit bug in the functionPlanets::ERA(const Time &timeUT1), according to the code:

Double Planets::ERA(const Time &timeUT1)
{
  const Time T = timeUT1-51544.5;
  return fmod(2*PI*(0.7790572732640 + T.mjdMod() + 0.00273781191135448*T.mjd()), 2*PI);
}

I guess it comes from the function of iauEra00 in SOFA library as shown below:
/* Days since fundamental epoch. */
   t = d1 + (d2- 2400000.5);

/* Fractional part of T (days). */
   **f = fmod(d1, 1.0) + fmod(d2, 1.0);**

/* Earth rotation angle at this UT1. */
   theta = iauAnp(D2PI * (f + 0.7790572732640 + 0.00273781191135448 * t));

Here d1 and d2 are two-part JD. However, as JD = MJD + 2400000.5, the correct expression is supposed to be:
fmod(2*PI*(0.7790572732640 + 0.5 + T.mjdMod() + 0.00273781191135448*T.mjd()), 2*PI);

**Comment by tmayerguerr** [comment]
Hi,

I think the code is correct
```  const Time T = timeUT1-mjd2time(J2000);  // with J2000=51544.5```
And in SOFA:
```  t = d1 + (d2- DJ00);  // with DJ00 = 2451545.0```

In SOFA the fractional part `f` is computed from `d1,d2` in JD whereas in GROOPS the fractional part it is computed from `T` (where 0.5 is already reduced).

regards
Torsten

**Comment by watercs** [comment]
Dear Torsten,

I hope this message finds you well. I wanted to extend my sincere appreciation for your quick response. I'd like to apologize for any oversight on my part when initially initiating the discussion.

My inquiry pertains to an issue I encountered while comparing the transformation matrix between ECI and ECEF using the SOFA library. Specifically, I observed a discrepancy in the calculation of the parameter '_f_,' while the value of 't' remains consistent with SOFA as you noted. Allow me to provide an illustrative example for clarity:

In the case of UT1 , d1 = 2454195.5 and d2 = 0.499999165813831, SOFA library calculates 'f' as:
_f = fmod(d1, 1.0) + fmod(d2, 1.0) = 0.99999165813831;_

However, when utilizing Groops, where 'mjdInt' equals 54195 and 'mjdMod' equals 0.499999165813831, we find that '_f_' is computed as:
_f = T.mjdMod() = 0.499999165813831._

I believe that this discrepancy can be attributed to the Earth's rotation starting from midnight (MJD begins at midnight, whereas JD starts at noon). Hope this would be helpful.

Lastly, I would like to take this opportunity to express my deep gratitude to you and your team for sharing the outstanding work of Groops online. Your dedication and contributions are greatly appreciated.

Best regards,
Changyong He

**Comment by tmayerguerr** [comment]
Dear Changyong He,

I think your thinking error is:
```
 timeUT1 = Time(54195, 0.499999165813831) // in MJD
 timeUT1.mjdMod() = 0.499999165813831
```
but T instead of timeUT1 is used:
```
Time T = timeUT1-mjd2time(J2000);   // with J2000=51544.5
T.mjdMod() = 0.99999165813831 // as wanted
```
regards
Torsten

**Reply by watercs** [reply]
Hi Torsten,

Thanks for your reply. Yes, you are right. The 0.5 is already reduced for T in GROOPS. 

Best regards,
Changyong
